File: rust/src/simplify_path.rs
//! This module was written to solve the "Simplify Path" problem from LeetCode.
//!
//! Given a string path, which is an absolute path (starting with a slash '/') to a file or directory in a Unix-style file system, convert it to the simplified canonical path.
//! 
//! In a Unix-style file system, a period '.' refers to the current directory, a double period '..' refers to the directory up a level, and any multiple consecutive slashes (i.e. '//') are treated as a single slash '/'. For this problem, any other format of periods such as '...' are treated as file/directory names.
//! 
//! The canonical path should have the following format:
//! 
//! - The path starts with a single slash '/'.
//! - Any two directories are separated by a single slash '/'.
//! - The path does not end with a trailing '/'.
//! - The path only contains the directories on the path from the root directory to the target file or directory (i.e., no period '.' or double period '..')
//! 
//! Return the simplified canonical path.
//! 
//! Examples:
//! Input: path = "/home/"
//! Output: "/home"
//! 
//! Input: path = "/home//foo/"
//! Output: "/home/foo"
//! 
//! Usage: simplify_path(path)

/// Simplifies an OS path, removing `..`, `.` and `//`
pub fn simplify_path(path: String) -> String {
  let paths = path.split('/').skip(1);

  let mut simplified_path: Vec<&str> = vec![];

  for element in paths {
      if element.is_empty() || element == "." {
        continue;
      }

      if element == ".." {
        simplified_path.pop();
        continue;
      }

      simplified_path.push(element);
  }

  format!("/{}", simplified_path.join("/"))
}

File: py/simplify_path.py
"""
Given an absolute path for a Unix-style file system, which begins with a slash '/', transform this path into its simplified canonical path.

In Unix-style file system context, a single period '.' signifies the current directory, a double period ".." denotes moving up one directory level, and multiple slashes such as "//" are interpreted as a single slash. In this problem, treat sequences of periods not covered by the previous rules (like "...") as valid names for files or directories.

The simplified canonical path should adhere to the following rules:

It must start with a single slash '/'.
Directories within the path should be separated by only one slash '/'.
It should not end with a slash '/', unless it's the root directory.
It should exclude any single or double periods used to denote current or parent directories.
Return the new path.

Examples:
Input: path = "/home/"
Output: "/home"

Input: path = "/home//foo/"
Output: "/home/foo"

Input: path = "/home/user/Documents/../Pictures"
Output: "/home/user/Pictures"

Input: path = "/../"
Output: "/"

Input: path = "/.../a/../b/c/../d/./"
Output: "/.../b/d"

Problem source: LeetCode

Usage: simplify_path.py <path>
"""

import sys

def simplify_path(path: str) -> str:
  directories = path.split("/")
  stack = []

  for directory in directories:
    if len(directory) == 0 or directory == "." or (directory == ".." and not stack):
      continue

    if directory == "..":
      stack.pop()
    else:
      stack.append(directory)

  return "/" + "/".join(stack)


if __name__ == "__main__":
  print(simplify_path(sys.argv[1]))